x-pack/plugins/task_manager/server/task_scheduling.ts

```diff
@@ -100,11 +113,58 @@ export class TaskScheduling {
    */
   public async runNow(taskId: string): Promise<RunNowResult> {
     return new Promise(async (resolve, reject) => {
-      this.awaitTaskRunResult(taskId).then(resolve).catch(reject);
+      this.awaitTaskRunResult(taskId)
+        // don't expose state on runNow
+        .then(({ id }) => resolve({ id }))
+        .catch(reject);
       this.taskPollingLifecycle.attemptToRun(taskId);
     });
   }
 
+  /**
+   * Run an ad-hoc task in memory without persisting it into ES or distributing the load across the cluster.
+   *
+   * @param task - The ephemeral task being queued.
+   * @returns {Promise<ConcreteTaskInstance>}
+   */
+  public async ephemeralRunNow(
+    task: EphemeralTask,
+    options?: Record<string, unknown>
+  ): Promise<RunNowResult> {
+    const id = uuid.v4();
+    const { taskInstance: modifiedTask } = await this.middleware.beforeSave({
+      ...options,
+      taskInstance: task,
+    });
+    return new Promise(async (resolve, reject) => {
+      const { cancel, resolveOnCancel } = cancellablePromise();
+      this.awaitTaskRunResult(id, resolveOnCancel)
+        .then((arg: RunNowResult) => {
+          resolve(arg);
+        })
+        .catch((err: Error) => {
+          reject(err);
+        });
+      const attemptToRunResult = this.ephemeralTaskLifecycle.attemptToRun({
```

Comment:
Good question!

The actual promise returned from this function is resolved _after_ the `awaitTaskRunResult` promise resolves. However, we do not wait to `await` this promise, as we want later execution to happen in parallel. The `awaitTaskRunResult` promise is resolved once the ephemeral task is successfully executed (technically, when a `TaskEventType.TASK_RUN` is emitted with the same `id`). 

However, the ephemeral task won't even get [into the queue](https://github.com/elastic/kibana/blob/b7ea9d98c1c3e07458cfdec967ebeb3463c2059a/x-pack/plugins/task_manager/server/ephemeral_task_lifecycle.ts#L192) until the subsequent `this.ephemeralTaskLifecycle.attemptToRun` is called (which puts it in the queue). 

The reason for all this confusion? Timing.

In the `this.ephemeralTaskLifecycle.attemptToRun`, it's possible that the ephemeral task is put into the queue and processed before this function call returns anything. If that happens, putting the `awaitTaskRunResult` _after_ would just hang because the task already completed. We need to listen for the completion _before_ we add it to the queue to avoid this possibility.

Does this address your question? I know it's a bit confusing but I think it's the best way, unless I'm misunderstanding what you're asking/suggesting?

Reply:
Thank you for the explanation. That make sense to me! Might be nice to leave this explanation as a comment for future devs who touch this code :)